src/Progress/Snapshot.php

```diff
@@ -4,26 +4,95 @@
 namespace BristolSU\Support\Progress;
 
 
+use BristolSU\Support\Activity\Activity;
 use BristolSU\Support\ActivityInstance\ActivityInstance;
 use BristolSU\Support\ActivityInstance\Contracts\ActivityInstanceRepository;
 use BristolSU\Support\ModuleInstance\Facade\ModuleInstanceEvaluator;
+use BristolSU\Support\Progress\ProgressUpdateRepository;
 use Carbon\Carbon;
+use Illuminate\Support\Facades\Hash;
 
 class Snapshot
 {
 
+    private ProgressUpdateRepository $ProgressUpdateRepository;
+
+    public function __construct()
+    {
+        $this->ProgressUpdateRepository = new ProgressUpdateRepository();
+    }
+
+    public function createActivityHashArray(Progress $activityProgress)
+    {
+        return [
+            $activityProgress->getActivityInstanceId(),
+            $activityProgress->getPercentage(),
+            json_encode(
+                array_map(fn(ModuleInstanceProgress $moduleInstanceProgress): array => [
+                    'moduleInstanceId' => $moduleInstanceProgress->getModuleInstanceId(),
+                    'mandatory' => $moduleInstanceProgress->isMandatory(),
+                    'complete' => $moduleInstanceProgress->isComplete(),
+                    'percentage' => $moduleInstanceProgress->getPercentage(),
+                    'active' => $moduleInstanceProgress->isActive(),
+                    'visible' => $moduleInstanceProgress->isVisible()
+                ], $activityProgress->getModules())
+            )
+        ];
+    }
+
+    public function ofUpdatesToActivity(Activity $activity, $caller)
+    {
+        $progresses = [];
+        foreach(app(ActivityInstanceRepository::class)->allForActivity($activity->id) as $activityInstance) {
+            array_push($progresses, $this->ofUpdateToActivityInstance($activityInstance, $caller));
+        }
+        return array_filter($progresses);
+    }
+
+    public function ofUpdateToActivityInstance(ActivityInstance $activityInstance, $caller) : ?Progress
+    {
+        // Set Cache key:
+        $itemKey = $caller . '_' . $activityInstance->id;
+
+        // Get the Current Progress:
+        $Progress = $this->ofActivityInstance($activityInstance);
+        $progressArrayHash = $this->createActivityHashArray($Progress);
```

Comment:
I'm not sure this class should know about the hash.

To keep responsibilities well defined, the repository should be able to determine how hashes are created.This way, if in the future someone has a custom solution for their storage and needs to specify a different kind of hash, we couldn't do this as the hash method can't be changed without breaking the original solution.

Reply:
That's fine, do you want multiple HashTypes on the Repository?

So a standard one that just Hashes an Array, and then an ActivityHash that process the above for example?

Reply:
Yeah could do - it doesn't really matter as this is all internals of the repository. Since we're using interfaces and abstracting single responsibilities I'm not sure there's much point in two hash functions, as they must only be used by the repository itself and the repository always hashes the same. I could see the point if other classes were allowed to use the hash functions, but since these have to be hidden either way works

Reply:
So just add a generateHash func which takes the createActivityHashArray from the Snapshot class? Do you think the Repo needs retitling to be ActivitiesProgressUpdateRepo?

Reply:
I think so - you can see what I'd expect the snapshot class to look like below. The snapshot class **does not** know anything about a hash, how it's created or why it's used. Therefore, the repository _public_ API shouldn't make mention of a hash ever. However, behind the scenes, the generateHash function can work the same as it does now in the snapshot class